core/src/main/scala/org/apache/spark/h2o/converters/H2ORDD.scala

```diff
@@ -58,63 +62,166 @@ class H2ORDD[A <: Product: TypeTag: ClassTag, T <: Frame] private(@transient val
   }
 
   /** Number of columns in the full dataset */
-  val numCols = frame.numCols()
-  val types = ReflectionUtils.types[A](colNames)
+  val numColsInFrame = frame.numCols()
+
+  val colNamesInFrame = frame.names()
+
+  val types = ReflectionUtils.types[A](colNamesInResult)
   val expectedTypesAll: Option[Array[Byte]] = ConverterUtils.prepareExpectedTypes(isExternalBackend, types)
 
   /**
    * :: DeveloperApi ::
    * Implemented by subclasses to compute a given partition.
    */
   override def compute(split: Partition, context: TaskContext): Iterator[A] = {
+    new H2ORDDIterator(frameKeyName, split.index)
+  }
 
-    val iterator = new H2OChunkIterator[A] {
-
-      val jc = implicitly[ClassTag[A]].runtimeClass
-      val cs = jc.getConstructors
-      val ccr = cs.collectFirst({
-                case c if c.getParameterTypes.length == colNames.length => c
-              })
-        .getOrElse({
-            throw new IllegalArgumentException(
-                  s"Constructor must take exactly ${colNames.length} args")
-      })
-
-      override val keyName = frameKeyName
-      override val partIndex = split.index
-      val expectedTypes: Option[Array[Byte]] = expectedTypesAll
-
-      /* Converter context */
-      override val converterCtx: ReadConverterContext =
-      ConverterUtils.getReadConverterContext(isExternalBackend,
-        keyName,
-        chksLocation,
-        expectedTypes,
-        partIndex)
+  private val columnTypeNames = ReflectionUtils.typeNames[A](colNamesInResult)
+
+  private val jc = implicitly[ClassManifest[A]].runtimeClass
+
+  private type Deserializer = ReadConverterContext => Int => Any
+
+  private val DefaultDeserializer: Deserializer = _ => _ => null
+
+  private val plainExtractors: Map[String, Deserializer] = Map(
+    "Boolean" -> ((source: ReadConverterContext) => (col: Int) => source.getBoolean(col)),
+    "Byte"    -> ((source: ReadConverterContext) => (col: Int) => source.getByte(col)),
+    "Double"  -> ((source: ReadConverterContext) => (col: Int) => source.getDouble(col)),
+    "Float"   -> ((source: ReadConverterContext) => (col: Int) => source.getFloat(col)),
+    "Integer" -> ((source: ReadConverterContext) => (col: Int) => source.getInt(col)),
+    "Long"    -> ((source: ReadConverterContext) => (col: Int) => source.getLong(col)),
+    "Short"   -> ((source: ReadConverterContext) => (col: Int) => source.getShort(col)),
+    "String"  -> ((source: ReadConverterContext) => (col: Int) => source.getString(col)))
+
+  private def returnOption[X](op: ReadConverterContext => Int => X) = (source: ReadConverterContext) => (col: Int) => opt(op(source)(col))
+
+  type TypeName = String
```

Comment:
The benefit is that it's much harder to mix two string parameters if they
have specific types.
Also, types may be turned into value classes, for better type safety (not
just token one).

Thanks,
-Vlad

On Fri, Aug 26, 2016 at 10:24 AM, Jakub Háva <notifications@github.com>
wrote:

> In core/src/main/scala/org/apache/spark/h2o/converters/H2ORDD.scala
> <https://github.com/h2oai/sparkling-water/pull/58#discussion_r76455352>:
>
> > +
> > +  private val DefaultDeserializer: Deserializer = _ => _ => null
> > +
> > +  private val plainExtractors: Map[String, Deserializer] = Map(
> > +    "Boolean" -> ((source: ReadConverterContext) => (col: Int) => source.getBoolean(col)),
> > +    "Byte"    -> ((source: ReadConverterContext) => (col: Int) => source.getByte(col)),
> > +    "Double"  -> ((source: ReadConverterContext) => (col: Int) => source.getDouble(col)),
> > +    "Float"   -> ((source: ReadConverterContext) => (col: Int) => source.getFloat(col)),
> > +    "Integer" -> ((source: ReadConverterContext) => (col: Int) => source.getInt(col)),
> > +    "Long"    -> ((source: ReadConverterContext) => (col: Int) => source.getLong(col)),
> > +    "Short"   -> ((source: ReadConverterContext) => (col: Int) => source.getShort(col)),
> > +    "String"  -> ((source: ReadConverterContext) => (col: Int) => source.getString(col)))
> > +
> > +  private def returnOption[X](op: ReadConverterContext => Int => X) = (source: ReadConverterContext) => (col: Int) => opt(op(source)(col))
> > +
> > +  type TypeName = String
>
> Yup, I read about in on stackoverflow, but still don't see the benefit of
> doing type TypeName = String instead of using just String
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/h2oai/sparkling-water/pull/58/files/dda8bacb29caf9d9ad724922e0bdf29e831b7ecb#r76455352>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAOuK7NDjyYOtQs6BMTDXh3UZ_EBYkTxks5qjyE8gaJpZM4Jsgi_>
> .
>